File: test/t_BayesNetAgrum_std.py
#! /usr/bin/env python

from openturns import *
from pyAgrum import *
from otagrum import *

try :
    # Create an empty BN
    myBN = BayesNet()
    # Create light variable
    light = LabelizedVar("Light", "Light", 0)
    light.addLabel("Dim")
    light.addLabel("Bright")
    lightSize = len(light)
    # Create moisture vriable
    moisture = LabelizedVar("Moisture", "Moisture", 0)
    moisture.addLabel("Dry")
    moisture.addLabel("Wet")
    moistureSize = len(moisture)
    # Create height variable
    height = DiscretizedVar("Height", "Height")
    data = NumericalPoint(range(10, 100, 10))
    heightWhenDimAndDry = Uniform(0.0, 40.0)
    heightWhenDimAndWet = Uniform(10.0, 60.0)
    heightWhenBrightAndDry = Uniform(0.0, 20.0)
    heightWhenBrightAndWet = Uniform(40.0, 100.0)
    data = BayesNetAgrum.AdaptGrid(DistributionCollection([heightWhenDimAndDry, heightWhenDimAndWet, heightWhenBrightAndDry, heightWhenBrightAndWet]), data)
    heightSize = data.getSize() - 1
    for i in list(data):
        height.addTick(i)
    # Add variables to the net
    indexLight    = myBN.add(light)
    indexMoisture = myBN.add(moisture)
    indexHeight   = myBN.add(height)
    # Create arcs
    myBN.insertArc(indexLight, indexMoisture)
    myBN.insertArc(indexLight, indexHeight)
    myBN.insertArc(indexMoisture, indexHeight)
    # Create conditional probability tables
    myBN.cpt(indexLight)[:]= [0.5, 0.5]
    myBN.cpt(indexMoisture)[{'Light' : 'Dim'}] = [0.2, 0.8]
    myBN.cpt(indexMoisture)[{'Light' : 'Bright'}] = [0.6, 0.4]
    myBN.cpt(indexHeight)[{'Light': 'Dim', 'Moisture': 'Dry'}]   = BayesNetAgrum().Discretize(heightWhenDimAndDry, data)
    myBN.cpt(indexHeight)[{'Light': 'Bright', 'Moisture': 'Dry'}] = BayesNetAgrum().Discretize(heightWhenBrightAndDry, data)
    myBN.cpt(indexHeight)[{'Light': 'Dim', 'Moisture': 'Wet'}]    = BayesNetAgrum().Discretize(heightWhenDimAndWet, data)
    myBN.cpt(indexHeight)[{'Light': 'Bright', 'Moisture': 'Wet'}] = BayesNetAgrum().Discretize(heightWhenBrightAndWet, data)
    # Create a BayesNetAgrum object
    otbn = BayesNetAgrum(myBN)
    # Export to BIF file
    otbn.exportToBIFFile("test.bif")
    # Perform inference
    for evidenceValue in ["Bright", "Dim"]:
        # First, set evidence
        otbn.setEvidence("Light", evidenceValue)
        # Second, get a marginal distribution of interrest
        heightDistribution = otbn.getMarginal("Height")
        heightDistribution.setName("Height_Light_" + evidenceValue)
        # Use the distribution the way you want
        print "heightDistribution=", heightDistribution
        heightDistribution.drawPDF(-10.0, 110.0, 512).draw("PDF_" + heightDistribution.getName())
        heightDistribution.drawCDF(-10.0, 110.0, 512).draw("CDF_" + heightDistribution.getName())
    
except :
    import sys
    print "t_BayesNetAgrum_std.py", sys.exc_type, sys.exc_value

File: test/t_BayesNetAgrum_std.cxx
#include <iostream>
// aGrUM includes
#include <agrum/BN/BayesNet.h>
#include <agrum/BN/io/BIF/BIFWriter.h>
#include <agrum/multidim/labelizedVariable.h>
#include <agrum/multidim/discretizedVariable.h>
#include <agrum/BN/inference/lazyPropagation.h>
// OT includes
#include <NumericalPoint.hxx>
#include <Collection.hxx>
#include <Uniform.hxx>
#include <Distribution.hxx>
// OTBayesNet
#include "BayesNetAgrum.hxx"

using namespace gum;
using namespace OT;
using namespace OT::Base::Type;
using namespace OT::Base::Stat;
using namespace OT::Uncertainty::Distribution;
using namespace OT::Uncertainty::Model;

typedef OT::Base::Type::Collection<Distribution>             DistributionCollection;

int main(int argc, char **argv)
{
  // Create an empty BN
  BayesNet<float> myBN;
  // Create variables
  LabelizedVariable light("Light", "Light", 0);
  LabelizedVariable moisture("Moisture", "Moisture", 0);
  DiscretizedVariable<float> height("Height", "Height");
  // Create labels and ticks
  light.addLabel("Dim");
  light.addLabel("Bright");
  int lightSize(light.domainSize());
  moisture.addLabel("Dry");
  moisture.addLabel("Wet");
  int moistureSize(moisture.domainSize());
  NumericalPoint data;
  for (UnsignedInteger i = 0; i <= 10; ++i)
    data.add(10.0 * i);
  int heightSize(data.getSize() - 1);
  for (int i = 0; i < heightSize + 1; ++i)
    height.addTick(data[i]);
  // Add variables to the net
  gum::Id indexLight(myBN.addVariable(light));
  gum::Id indexMoisture(myBN.addVariable(moisture));
  gum::Id indexHeight(myBN.addVariable(height));
  // Create arcs
  myBN.insertArc(indexLight, indexMoisture);
  myBN.insertArc(indexLight, indexHeight);
  myBN.insertArc(indexMoisture, indexHeight);
  // Create conditional probability tables
  {
    const Potential<float>& probaLight(myBN.cpt(indexLight));
    const int n = lightSize;
    const float t[] = {0.5, 0.5};
    const std::vector<float> v(t, t + n);
    probaLight.fillWith(v);
  }
  {
    const Potential<float>& probaMoisture(myBN.cpt(indexMoisture));
    const int n = lightSize * moistureSize;
    const float t[] = {0.2, 0.8,
                       0.6, 0.4};
    const std::vector<float> v(t, t + n);
    probaMoisture.fillWith(v);
  }
  {
    const Potential<float>& probaHeight(myBN.cpt(indexHeight));
    Uniform heightWhenDimAndDry(0.0, 40.0);
    Uniform heightWhenBrightAndDry(0.0, 20.0);
    Uniform heightWhenDimAndWet(10.0, 60.0);
    Uniform heightWhenBrightAndWet(40.0, 100.0);
    BayesNetAgrum::DistributionCollection collection;
    collection.add(heightWhenDimAndDry);
    collection.add(heightWhenBrightAndDry);
    collection.add(heightWhenDimAndWet);
    collection.add(heightWhenBrightAndWet);
    probaHeight.fillWith(BayesNetAgrum::Discretize(collection, data));
  }
  BayesNetAgrum otbn(myBN);
  std::cout << "otbn=" << otbn << std::endl;
  otbn.exportToBIFFile("test.bif");
  std::cout << "Evidence: Light=Bright" << std::endl;
  otbn.setEvidence("Light", "Bright");
  Distribution heightDistribution(otbn.getMarginal("Height"));
  std::cout << "heightDistribution=" << heightDistribution << std::endl;
  std::cout << "proba(height < 30.0)=" << heightDistribution.computeCDF(30.0) << std::endl;
  std::cout << "Evidence: Light=Dim" << std::endl;
  otbn.setEvidence("Light", "Dim");
  heightDistribution = otbn.getMarginal("Height");
  std::cout << "heightDistribution=" << heightDistribution << std::endl;
  std::cout << "proba(height < 30.0)=" << heightDistribution.computeCDF(30.0) << std::endl;
  return 0;
}
